### PR TITLE
Fixes #21384 - Support Puppet 4 to Puppet 5 upgrade

### PR DESCRIPTION
**Pull request by ehelms**

### CODE REVIEW

**Comment by ehelms** [comment]
This needs some testing, as I hit odd issues with the puppetversion fact resolving. To test this:

 1) vagrant up centos7-katello-nightly (or 3.5)
 2) apply patch to /usr/share/katello-installer-base/hooks
 3) install puppet 5 release RPM https://yum.puppetlabs.com/puppet5/el/7/x86_64/puppet-release-1.0.0-1.el7.noarch.rpm
  4) foreman-installer --ugprade-puppet

**Comment by chris1984** [comment]
Code looks fine.

**Comment by ekohl** [comment]
I suspect that you want to reinstall the oauth gem package. The Puppet 4 collection is Ruby 2.1.9 and Puppet 5 is 2.4.x and we have an ugly post section in the RPM to install it into the Puppet collection.

**Comment by chris1984** [comment]
Reinstalling the following package made the installer have a good run:

  puppet-agent-oauth.noarch 0:0.5.1-3.el7                                                                                  

Complete!
[root@centos7-katello-nightly ~]# foreman-installer --disable-system-checks
Installing             Puppet_authorization_hocon_rule[rule-puppetlabs cs [66%] [...........................................                       Installing             Done                                               [100%] [................................................................]
  Success!
  * Katello is running at https://centos7-katello-nightly.vault111.example.com
  * To install an additional Foreman proxy on separate machine continue by running:

      foreman-proxy-certs-generate --foreman-proxy-fqdn "$FOREMAN_PROXY" --certs-tar "/root/$FOREMAN_PROXY-certs.tar"
  The full log is at /var/log/foreman-installer/katello.log

**Comment by chris1984** [comment]
Looks like its also hanging on 

/usr/share/katello-installer-base/hooks/post/31-upgrade-puppet.rb:24:in `block (4 levels) in load'^C: undefined method `puppet5_installed?' for #<Kafo::HookContext:0x00000001340b60> (NoMethodError)

after a normal install run

**Comment by ehelms** [comment]
Updated, ready for re-review

**Comment by chris1984** [comment]
Currently its hanging at:

================================================================================
 Package                  Arch         Version              Repository     Size
================================================================================
Reinstalling:
 puppet-agent-oauth       noarch       0.5.1-3.el7          foreman        45 k

Transaction Summary
================================================================================
Reinstall  1 Package

Total download size: 45 k
Installed size: 42 k

The cpu is spiking to 150% and the puppet server logs keep repeating with:

2017-10-20 21:50:06,931 ERROR [main] [p.t.internal] Error during app buildup!
java.lang.RuntimeException: Service ':FilesystemWatchService' not found
	at puppetlabs.trapperkeeper.internal$handle_prismatic_exception_BANG_.invokeStatic(internal.clj:121)
	at puppetlabs.trapperkeeper.internal$handle_prismatic_exception_BANG_.invoke(internal.clj:111)
	at puppetlabs.trapperkeeper.internal$instantiate$fn__13800.invoke(internal.clj:162)
	at puppetlabs.trapperkeeper.internal$instantiate.invokeStatic(internal.clj:159)
	at puppetlabs.trapperkeeper.internal$instantiate.invoke(internal.clj:152)
	at puppetlabs.trapperkeeper.internal$eval14334$build_app_STAR___14343$fn__14344.invoke(internal.clj:569)
	at puppetlabs.trapperkeeper.internal$eval14334$build_app_STAR___14343.invoke(internal.clj:537)
	at puppetlabs.trapperkeeper.internal$eval14436$boot_services_STAR___14445$fn__14446$fn__14447.invoke(internal.clj:639)
	at puppetlabs.trapperkeeper.internal$eval14436$boot_services_STAR___14445$fn__14446.invoke(internal.clj:638)
	at puppetlabs.trapperkeeper.internal$eval14436$boot_services_STAR___14445.invoke(internal.clj:633)
	at puppetlabs.trapperkeeper.core$eval15266$boot_with_cli_data__15273$fn__15274.invoke(core.clj:130)
	at puppetlabs.trapperkeeper.core$eval15266$boot_with_cli_data__15273.invoke(core.clj:95)
	at puppetlabs.trapperkeeper.core$eval15295$run__15300$fn__15301.invoke(core.clj:151)
	at puppetlabs.trapperkeeper.core$eval15295$run__15300.invoke(core.clj:145)
	at puppetlabs.trapperkeeper.core$main.invokeStatic(core.clj:173)
	at puppetlabs.trapperkeeper.core$main.doInvoke(core.clj:159)
	at clojure.lang.RestFn.invoke(RestFn.java:512)
	at clojure.lang.Var.invoke(Var.java:409)
	at clojure.lang.AFn.applyToHelper(AFn.java:178)
	at clojure.lang.Var.applyTo(Var.java:700)
	at clojure.core$apply.invokeStatic(core.clj:646)
	at clojure.core$apply.invoke(core.clj:641)
	at puppetlabs.trapperkeeper.main$_main.invokeStatic(main.clj:7)
	at puppetlabs.trapperkeeper.main$_main.doInvoke(main.clj:4)
	at clojure.lang.RestFn.invoke(RestFn.java:512)
	at clojure.lang.Var.invoke(Var.java:409)
	at clojure.lang.AFn.applyToHelper(AFn.java:178)
	at clojure.lang.Var.applyTo(Var.java:700)
	at clojure.core$apply.invokeStatic(core.clj:646)
	at clojure.main$main_opt.invokeStatic(main.clj:314)
	at clojure.main$main_opt.invoke(main.clj:310)
	at clojure.main$main.invokeStatic(main.clj:421)
	at clojure.main$main.doInvoke(main.clj:384)
	at clojure.lang.RestFn.invoke(RestFn.java:619)
	at clojure.lang.Var.invoke(Var.java:430)
	at clojure.lang.AFn.applyToHelper(AFn.java:195)
	at clojure.lang.Var.applyTo(Var.java:700)
	at clojure.main.main(main.java:37)
2017-10-20 21:50:06,954 INFO  [Thread-2] [p.t.internal] Shutting down due to JVM shutdown hook.
2017-10-20 21:50:06,960 INFO  [Thread-2] [p.t.internal] Beginning shutdown sequence
2017-10-20 21:50:06,972 INFO  [Thread-2] [p.t.internal] Finished shutdown sequence

Will try with more ram then the 6gb we give the vagrant boxes and see what happens.

File: hooks/init/31-upgrade-puppet.rb [hunk 1]
```diff
@@ -1,34 +1,24 @@
...
-def copy_data
-  success = []
-  success << Kafo::Helpers.execute('cp -rfp /etc/puppet/environments/* /etc/puppetlabs/code/environments') if File.directory?('/etc/puppet/environments') && !Dir['/etc/puppet/environments/*'].empty?
-  success << Kafo::Helpers.execute('cp -rfp /var/lib/puppet/ssl /etc/puppetlabs/puppet') if File.directory?('/var/lib/puppet/ssl')
-  success << Kafo::Helpers.execute('cp -rfp /var/lib/puppet/foreman_cache_data /opt/puppetlabs/puppet/cache/') if File.directory?('/var/lib/puppet/foreman_cache_data')
-  !success.include?(false)
+def upgrade_puppet_package
+  Kafo::Helpers.execute('yum upgrade -y puppetserver')
+  Kafo::Helpers.execute('yum upgrade -y puppet-agent')
+  Kafo::Helpers.execute('yum reinstall puppet-agent-oauth')
```
Comment (chris1984) [inline hooks/init/31-upgrade-puppet.rb]: Stopping the service and then reinstalling the package made it go through, can we do a stop, upgrade/reinstall then start back up?

Comment (ehelms) [reply hooks/init/31-upgrade-puppet.rb]: Ahh yea, you are entirely right. I shoulda put the stop services that is there before upgrading the packages. This has been updated with that switch but I haven't tested it.

**Comment by ekohl** [comment]
Should it also reset the ```puppet_puppetserver_version``` answer?

**Comment by ehelms** [comment]
@ekohl It is, probably just buried in the diff -- https://github.com/Katello/katello-installer/pull/557/files#diff-c63eafe79c5bc8b032ccbd0e6effe674R19

**Comment by ekohl** [comment]
That should teach me not to do reviews in the morning.

**Comment by ehelms** [comment]
@chris1984 This is ready for retesting

**Comment by chris1984** [comment]
Ack, will test when I get back from my customer visit today.

**Comment by chris1984** [comment]
@ehelms Still hanging, I let it sit for a few mins incase it didnt have enough ram/cpu :( Tested on 3.5, nightly said puppet was already upgraded.

~~~
Upgrade Step: upgrade_puppet_package...
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
 * base: mirror.linux.duke.edu
 * epel: download-ib01.fedoraproject.org
 * extras: repo.us.bigstepcloud.com
 * updates: mirror.jax.hugeserver.com
No packages marked for update
yum upgrade -y puppetserver finished successfully!
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
 * base: mirror.linux.duke.edu
 * epel: download-ib01.fedoraproject.org
 * extras: repo.us.bigstepcloud.com
 * updates: mirror.jax.hugeserver.com
No packages marked for update
yum upgrade -y puppet-agent finished successfully!
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
 * base: mirror.linux.duke.edu
 * epel: download-ib01.fedoraproject.org
 * extras: repo.us.bigstepcloud.com
 * updates: mirror.jax.hugeserver.com
Resolving Dependencies
--> Running transaction check
---> Package puppet-agent-oauth.noarch 0:0.5.1-3.el7 will be reinstalled
--> Finished Dependency Resolution

Dependencies Resolved

================================================================================
 Package                  Arch         Version              Repository     Size
================================================================================
Reinstalling:
 puppet-agent-oauth       noarch       0.5.1-3.el7          foreman        45 k

Transaction Summary
================================================================================
Reinstall  1 Package

Total download size: 45 k
Installed size: 42 k
~~~

**Comment by ekohl** [comment]
@chris1984 what are the specs of the system in question?

**Comment by chris1984** [comment]
@ekohl just the standard vagrant nighly box 
~~~
MemTotal:        5946132 kB
~~~
with 2 cpus

File: hooks/init/31-upgrade-puppet.rb [hunk 2]
```diff
@@ -1,34 +1,24 @@
...
```
Comment (chris1984) [inline hooks/init/31-upgrade-puppet.rb]: Got it to go through this reinstall needs to be with a -y so its not hung waiting for input.

**Comment by ehelms** [comment]
Updated

**Comment by chris1984** [comment]
Looks good, thanks @ehelms 

~~~
Puppet 4 to 5 upgrade initialization complete, continuing with installation
Resetting puppet params...
Puppet 4 to 5 upgrade param reset, continuing with installation
Installing             Done                                               [100%] [.......................................................................................................................................................................................................]
  Success!
  * Katello is running at https://centos7-katello-nightly.vault111.example.com
  * To install an additional Foreman proxy on separate machine continue by running:

      foreman-proxy-certs-generate --foreman-proxy-fqdn "$FOREMAN_PROXY" --certs-tar "/root/$FOREMAN_PROXY-certs.tar"
  The full log is at /var/log/foreman-installer/katello.log
~~~

**Pull request merged**